src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -2327,125 +2328,128 @@ int SketchObject::delConstraintOnPoint(int VertexId, bool onlyCoincident)
+        return true;
+    };
+
+    auto performAllConstraintChecksOrChanges = [&](auto& constr) -> std::optional<bool> {
+        if (constr->Type != Sketcher::Coincident && onlyCoincident) {
+            return true;
+        }
+        switch (constr->Type) {
+            case Sketcher::Coincident:
+                return performCoincidenceChecksOrChanges(constr);
+            case Sketcher::Distance:
+            case Sketcher::DistanceX:
+            case Sketcher::DistanceY: {
+                return (transferToReplacement(constr->First, constr->FirstPos)
+                        || transferToReplacement(constr->Second, constr->SecondPos));
+            }
+            case Sketcher::PointOnObject: {
+                return transferToReplacement(constr->First, constr->FirstPos);
+            }
+            case Sketcher::Tangent:
+            case Sketcher::Perpendicular: {
+                // we could keep this constraint by converting it to a simple one, but that doesn't
+                // always work (for example if tangent-via-point is necessary), and it is not really
+                // worth it
+                return false;
             }
-            else if ((*it)->Type == Sketcher::Symmetric) {
-                if (((*it)->First == GeoId && (*it)->FirstPos == PosId)
-                    || ((*it)->Second == GeoId && (*it)->SecondPos == PosId)) {
-                    continue;// skip this constraint
-                }
+            case Sketcher::Vertical:
+            case Sketcher::Horizontal:
+            case Sketcher::Symmetric: {
+                return false;
             }
-            else if ((*it)->Type == Sketcher::Vertical || (*it)->Type == Sketcher::Horizontal) {
-                if (((*it)->First == GeoId && (*it)->FirstPos == PosId)
-                    || ((*it)->Second == GeoId && (*it)->SecondPos == PosId)) {
-                    continue;// skip this constraint
-                }
+            default:
+                return std::nullopt;
+        }
+    };
+
+    // remove or redirect any constraints associated with the given point
+    for (auto& constr : vals) {
+        // keep the constraint if it doesn't involve the point
+        if (!constr->involvesGeoIdAndPosId(geoId, posId)) {
+            // for these constraints remove the constraint even if it is not directly associated
+            // with the given point
+            if ((constr->Type == Sketcher::Distance || constr->Type == Sketcher::DistanceX
+                 || constr->Type == Sketcher::DistanceY)
+                && (constr->First == geoId && constr->FirstPos == PointPos::none
+                    && (posId == PointPos::start || posId == PointPos::end))) {
+                continue;
             }
+            newVals.push_back(constr);
+            continue;
+        }
+        if (performAllConstraintChecksOrChanges(constr) != false) {
+            newVals.push_back(constr);
         }

```

Comment:
I don't see why checking if ther return value from the function is not `false`, removing ` != false` should give the same result.
```c++
        if (performAllConstraintChecksOrChanges(constr)) {
            newVals.push_back(constr);
        }
```

Reply:
Oh I now understand, thanks. But I think that's a bit confusing.

Also, the dual purpose makes it a bit difficult to follow, as the reader has to account for all cases `std::nullopt`/`true`/`false`.

While it works, it's a bit too smart for my liking.
I think I would have written it something like this instead:

```c++
if (performAllConstraintChecksOrChanges(constr).value_or(true)) {
    // Keep the constraint unless explicitly asked to delete
    newVals.push_back(constr);
}
```

Reply:
What do you mean "dual purpose"?

Also, didn't know about (rather didn't bother to remember) `value_or`. I guess it becomes "a bit to smart for me" now :D.

Reply:
I meant that the check both checks if there is a value _and_ the value of the boolean.
By using a default value like in `value_or` solves that issue :)

Reply:
changed to `...value_or...` locally.